### DISCUSSION TITLE
WearOS 3.5 isn't re-establishing connection with TicWatch using MiaoMiao 1,Libre 2 and OOP2

### DISCUSSION THREAD

**Discussion by SimonPhilpott** [NightscoutFoundation/xDrip#3272] [Bug reports]
I've been working through this issue for a while, looking through all of the settings available to me and I still havent resolved this.

 xDrip version (classic status page)
_The latest stable (July 2023), also tested on the latest nightly. 1d84ed1-2023.12.23_

The brand and model of your phone, what you have on the hardware data source page, 
_Samsung Note 10+, TicWatch Pro 3 Ultra, MiaoMiao 1, Libre 2 (EU)_

_The steps you took that led to the problem._

This problem has become a lot more prevailent since my Ticwatch moved to Android Wear OS 3.5. This forced me to move away from the WearOS app on my phone to using Mobovi Health.

This happend perhaps once a day previoisuly, whereas now I have to force the reconnection by restarting the collector almost once every couple of hours. I use my ticwatch pro 3 as my collector, not my phone. I use OOP2 alongside the Xdrip+ wearOS app to collect from my Libre 2 via my MiaoMiao 1;

- I have ensured that all apps including the new Mobovi Health app and Bluetooth are all whitelisted or excluded from battery opimisation. 
- I have turned on aggressive service restarts which should resolve my problem, but doesn't;t seem to restart the collector even after 20mins+ of the disconnected status of my collector is shown on the status screen.
- I am able to re-establish connection and get readings the first time once I manually press restart collector. But the connection is sometimes disconnected after anywhere up to an hour from then and doesn't automatically reconnect. 
- Since I use my watch as the collector, I have checked to ensure no battery optimisation is made on there for Xdrip also
- In my event log I see getLaterstBgForXMinutes: factor is 0 returning.
- I can see the messages, Aggressively restarting the collector service due to lack of reception: backoff 180 (and 360)
- I also see wearDexCollectionService : Watchdog triggered, attempting to reset Bluetooth (might this be the problem?, I'd rather rely on the aggressive collection restart than to reset the bluetooth?)

**Comment by Navid200** [comment]
Currently, no one is addressing WearOS.  
Not everyone knows everything.  For example, I have 0 experience with WearOS.  
Even if one does, they may not be interested in a particular subject.

As a result, there has been no update to the WearOS portion of xDrip for almost a year now.

Anyone interested in contributing is welcome to open a PR or a discussion to ask questions or to make suggestions.

**Reply by SimonPhilpott** [reply]
Yikes. Okay, I appreciate the honest answer. In Android Studio do you happen to know where the WearOS app is in the app structure? I've got some pretty rounded experience with Object O if I can familiarise myself within a personal repo I can set up some bespoke logging and see if I can single out was is, or isn't happening.

**Reply by Navid200** [reply]
That is a very good question.
I am a junior developer.  So, take the following with a grain of salt.

WearOS has been implemented in xDrip very inefficiently.  There is so much redundant code in xDrip for it.
When a developer adds a feature or edits an existing feature, they need to edit the identical code to also edit the WearOS portion.  But, not every developer uses WearOS.  So, as an example, even if I edit the WearOS portion, I will not be able to test it because I don't even have a WearOS device.  Therefore, every time I open a submission to xDrip, I only edit the main section and leave the WearOS portion alone.  As a result, there is a lot of discrepancy between the two portions and there is no clean easy way to keep the two in sync.

Now, to answer your question, please see the attached image.
![Screenshot 2024-01-04 143500](https://github.com/NightscoutFoundation/xDrip/assets/51497406/7dac5fc6-6bed-4949-b40e-e46ba9fc86fd)

**Reply by SimonPhilpott** [reply]
Right 👍, appreciated. It's been a long time since I contributed so I'll have to get back up to speed.

It's a shame that GitHub doesn't work better for open development. Even a basic Agile-like dashboard would flag what's under-resourced. As a repository it's great, but far flug developers aren't a team that set up calls and all that jazz.

**Reply by Navid200** [reply]
This is just a personal thought:

I wish someone would create a companion app for WearOS.  If that happens, the person who owns that companion app will have a lot of flexibility in customizing the WearOS support of xDrip.  For example, they will be able to add new features without needing to get approval from the xDrip team.
All WearOS related issues and feature request will be transferred to the companion app team.
The users of WearOS will benefit because their questions and requests will go directly to those with experience supporting WearOS.

To me, that is a win-win-win scenario!

We used to have a lot of open issues with MiBand.  Since MiBand support moved to a companion app, MiBand support has expanded and MiBand users are happy and we (xDrip) have no open issues related to MiBand.

**Reply by SimonPhilpott** [reply]
I guess that the second option would be to move towards encompassing the collection features of Xdrip into AAPS. It already has 90% of the features and is a standalone WearOS app already.
